processing/src/main/java/org/apache/druid/query/scan/ScanQueryRunnerFactory.java

```diff
@@ -68,34 +81,117 @@ public ScanQueryRunnerFactory(
+        // Use priority queue strategy
+        return sortAndLimitScanResultValues(
+            Sequences.concat(Sequences.map(
+                Sequences.simple(queryRunners),
+                input -> input.run(queryPlus, responseContext)
+            )),
+            query
+        );
+      } else if (numSegments <= scanQueryConfig.getMaxSegmentsTimeOrderedInMemory()) {
+        // Use n-way merge strategy
+        final Sequence<ScanResultValue> unbatched =
+            Sequences.map(
+                Sequences.simple(queryRunners),
+                (input) -> Sequences.concat(
+                    Sequences.map(
+                        input.run(queryPlus, responseContext),
+                        srv -> Sequences.simple(srv.toSingleEventScanResultValues())
+                    )
+                )
+            ).flatMerge(
+                seq -> seq,
+                Ordering.from(new ScanResultValueTimestampComparator(
+                    query
+                )).reverse() // This needs to be reversed because
+            ).limit(
+                Math.toIntExact(query.getLimit())
+            );
+
+        return unbatched;
       }
+      throw new UOE(
+          "Time ordering for queries of %,d segments per historical and a row limit of %,d is not supported."
+          + "  Try reducing the scope of the query to scan fewer segments than the configurable segment limit of"
+          + " %,d segments or lower the row limit below %,d.",
+          numSegments,
+          query.getLimit(),
+          scanQueryConfig.getMaxSegmentsTimeOrderedInMemory(),
+          scanQueryConfig.getMaxRowsQueuedForTimeOrdering()
+      );
     };
   }
 
+  @VisibleForTesting
+  Sequence<ScanResultValue> sortAndLimitScanResultValues(
+      Sequence<ScanResultValue> inputSequence,
+      ScanQuery scanQuery
+  )
+  {
+    Comparator<ScanResultValue> priorityQComparator = new ScanResultValueTimestampComparator(scanQuery);
+
+    // Converting the limit from long to int could theoretically throw an ArithmeticException but this branch
+    // only runs if limit < MAX_LIMIT_FOR_IN_MEMORY_TIME_ORDERING (which should be < Integer.MAX_VALUE)
+    int limit = Math.toIntExact(scanQuery.getLimit());
+
+    PriorityQueue<ScanResultValue> q = new PriorityQueue<>(limit, priorityQComparator);
+
+    Yielder<ScanResultValue> yielder = inputSequence.toYielder(
+        null,
+        new YieldingAccumulator<ScanResultValue, ScanResultValue>()
+        {
+          @Override
+          public ScanResultValue accumulate(ScanResultValue accumulated, ScanResultValue in)
+          {
+            yield();
+            return in;
+          }
+        }
+    );
+    while (!yielder.isDone()) {
+      ScanResultValue next = yielder.get();
+      List<ScanResultValue> singleEventScanResultValues = next.toSingleEventScanResultValues();
+      for (ScanResultValue srv : singleEventScanResultValues) {
```

Comment:
Is this also applicable for the none order case (i.e. the existing scan query implementation)?  Currently, there's no such limiting.

Reply:
^^Done.

Reply:
Yeah, I think it could be added to that, it would cut down on the result size from each historical and there aren't guarantees on what records are returned in the unordered limited case

Reply:
I think just using the Sequence#limit() operator will suffice for the non-ordered case:
```
        return Sequences.concat(
            Sequences.map(
                Sequences.simple(queryRunners),
                input -> input.run(queryPlus, responseContext)
            )
        ).limit(Math.toIntExact(query.getLimit()));
```

Reply:
I would make it conditional (maybe you have more than Int.MAX_VALUE rows?)